### PR TITLE
[AD] First pass at ECS listener

### PR DESCRIPTION
**Pull request by hkaj**
### What does this PR do?

Add support for fargate-based ECS tasks to auto discovery. Specifically:
- add an ecs service listener that polls regularly the ECS API for info about containers in the task.
- add an ecs config provider based on container labels extracted from tasks.
- add a tagger provider for ecs-related tags.
- create a container provider that can provide ecs info to the process agent for the live container view.
- refactor some utils that could be shared among listeners/providers.
- generally improves logging.

### Motivation

What inspired you to submit this pull request?

### Additional Notes

- The container struct is not great, containers can't be shared among runtimes/providers (ECS containers can't be treated as native docker containers in the agent right now). We should consider moving to an interface instead of a struct for the Container type def.
- Still need to take care of the hostname logic around ECS. Right now the hostname is the container id. Quite useless.
- I had to disable a log agent test that was flaky.

### CODE REVIEW

File: pkg/util/ecs/common.go [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+	var containers []*docker.Container
+	var stats ContainerStats
+
+	ecsContainers, err := GetECSContainers()
+	if err != nil {
+		log.Error("unable to get the container list from ecs")
+		return containers, err
+	}
+	for _, c := range ecsContainers {
+		entityID := fmt.Sprintf("docker://%s", c.DockerID)
```
Comment (mfpierre) [inline pkg/util/ecs/common.go]: use `ContainerIDToEntityName`

File: pkg/util/ecs/common.go [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+	resp, err := client.Get(StatsURL + "/" + c.DockerID)
+	if err != nil {
+		return stats, err
+	}
+	defer resp.Body.Close()
+
+	decoder := json.NewDecoder(resp.Body)
+	err = decoder.Decode(&stats)
+	if err != nil {
+		log.Errorf("decoding failed!") // TODO: delete me
```
Comment (mfpierre) [inline pkg/util/ecs/common.go]: delete?

File: pkg/util/container/providers.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+// ContainerProvider is the interface for singleton utils that can collect
+// containers (dockerutil, ecs, etc.)
+//
+// The goal of this interface is to provide a single method to get all containers
+// on the host without caring about the environment.
+// type ContainerProvider interface {
+// 	// TODO: we should not rely on docker types here
+// 	Containers(ContainerListConfig) ([]Container, error)
+// 	String() string
+// }
```
Comment (mfpierre) [inline pkg/util/container/providers.go]: cleanup commented code ⬆️

File: pkg/collector/listeners/ecs.go [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+func init() {
+	Register("ecs", NewECSListener)
+}
+
+// NewECSListener creates an ECSListener
+func NewECSListener() (ServiceListener, error) {
+	return &ECSListener{
+		services: make(map[string]Service),
+		stop:     make(chan bool),
+		t:        time.NewTicker(2 * time.Second),
```
Comment (mfpierre) [inline pkg/collector/listeners/ecs.go]: As checks run every 15 secs, 2 sec for AD seems a little short

File: pkg/collector/listeners/ecs.go [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+// refreshServices queries the task metadata endpoint for fresh info
+// compares the container list to the local cache and sends new/dead services
+// over newService and delService accordingly
+func (l *ECSListener) refreshServices() {
+	meta, err := ecs.GetTaskMetadata()
+	if err != nil {
+		log.Errorf("failed to get task metadata, not refreshing services - %s", err)
+		return
+	} else if meta.KnownStatus != "RUNNING" {
+		log.Debugf("task %s is not in RUNNING state yet, not refreshing services", meta.Family)
```
Comment (mfpierre) [inline pkg/collector/listeners/ecs.go]: return there?

File: pkg/collector/listeners/ecs.go [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+// ECSListener implements the ServiceListener interface for fargate-backed ECS cluster.
+// It pulls its tasks container list periodically and checks for
+// new containers to monitor, and old containers to stop monitoring
+type ECSListener struct {
+	task       ecs.TaskMetadata
+	services   map[string]Service // maps container IDs to services
+	newService chan<- Service
+	delService chan<- Service
+	stop       chan bool
+	m          sync.RWMutex
```
Comment (mfpierre) [inline pkg/collector/listeners/ecs.go]: you're not using the mutex when modifying the `services` map

File: pkg/tagger/collectors/ecs_fargate_main.go [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+		c.lastExpire = time.Now()
+		c.expireFreq = ecsFargateExpireFreq
+
+		c.expire, err = taggerutil.NewExpire(ecsFargateExpireFreq)
+
+		if err != nil {
+			return FetchOnlyCollection, fmt.Errorf("Failed to instantiate the container expiring process")
+		}
+		return FetchOnlyCollection, nil
+	} else {
```
Comment (mfpierre) [inline pkg/tagger/collectors/ecs_fargate_main.go]: useless `else` because of `return` in `if`

File: pkg/tagger/collectors/ecs_fargate_main.go [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+	}
+
+	expiries, err := c.parseExpires(deadCo)
+	if err != nil {
+		return err
+	}
+	c.infoOut <- expiries
+	c.lastExpire = time.Now()
+	return nil
+
```
Comment (mfpierre) [inline pkg/tagger/collectors/ecs_fargate_main.go]: remove extra line

**Review by mfpierre — DISMISSED** [review]
Looks good overall, a few things you might wanna address before merging though

File: pkg/collector/listeners/ecs.go (hunk 4 above)
Comment (hkaj) [reply pkg/collector/listeners/ecs.go]: For checks you're right. However the live container view also relies on it.

File: pkg/tagger/collectors/ecs_fargate_extract.go [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+			imageName := strings.Join(imageSplit[:len(imageSplit)-1], ":")
+			tags.AddLow("imageName", imageName)
+			if len(imageSplit) > 1 {
+				imageTag := imageSplit[len(imageSplit)-1]
+				tags.AddLow("image_tag", imageTag)
+			}
+
+			// container labels
+			for k, v := range ctr.Labels {
+				if _, found := labelBlackList[k]; !found {
```
[the remark this answers is not in the document]
Comment (hkaj) [reply pkg/tagger/collectors/ecs_fargate_extract.go]: i love this idea!

File: pkg/collector/listeners/ecs.go (hunk 5 above)
Comment (hkaj) [reply pkg/collector/listeners/ecs.go]: done

File: pkg/collector/listeners/ecs.go (hunk 6 above)
Comment (hkaj) [reply pkg/collector/listeners/ecs.go]: should be good now

File: pkg/tagger/collectors/ecs_fargate_extract.go [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+	for _, ctr := range meta.Containers {
+		seen[ctr.DockerID] = nil
+		if _, found := c.lastSeen[ctr.DockerID]; !found {
+			tags := utils.NewTagList()
+
+			// cluster
+			tags.AddLow("ecs_cluster_name", meta.ClusterName)
+
+			// task
+			tags.AddLow("ecs_task_family", meta.Family)
```
Comment (sunhay) [inline pkg/tagger/collectors/ecs_fargate_extract.go]: Hi! We're looking forward to using these tagger additions in the process agent :)

At a quick glance, I've noticed that these tag names are inconsistent with those from the other extractors; is there any way we can change them to match whats in `{docker,ecs}_extract.go`?

e.g:
`"ecs_task_family" -> "task_name"`, 
`"ecs_task_version" -> "task_version"`, 
`"imageName" -> "image_name"`, etc.

File: pkg/tagger/collectors/ecs_fargate_extract.go [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+			// cluster
+			tags.AddLow("cluster_name", meta.ClusterName)
+
+			// task
+			tags.AddLow("task_family", meta.Family)
+			tags.AddLow("task_version", meta.Version)
+
+			// container
+			tags.AddLow("ecs_container_name", ctr.Name)
+			tags.AddHigh("docker_container_name", ctr.DockerName)
```
Comment (sunhay) [inline pkg/tagger/collectors/ecs_fargate_extract.go]: `"docker_container_name"` -> `"container_name"`

Last one that I see that should change for consistency! 👌

File: pkg/util/container/providers.go [hunk 12]
```diff
@@ -0,0 +1,71 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2017 Datadog, Inc.
+
```
Comment (sunhay) [inline pkg/util/container/providers.go]: This also needs a `docker` build tag, I think. Also may/may not need to be stubbed if the `docker` tag isn't included?

Comment (hkaj) [reply pkg/util/container/providers.go]: you're right 👍 https://github.com/DataDog/datadog-agent/pull/1034 will fix this

File: pkg/util/ecs/common.go [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+// GetContainers returns all containers exposed by the ECS API
+// after transforming them into "generic" Docker containers.
+// TODO: add a cache
+func GetContainers() ([]*docker.Container, error) {
+	var containers []*docker.Container
+	var stats ContainerStats
+
+	ecsContainers, err := GetECSContainers()
+	if err != nil {
+		log.Error("unable to get the container list from ecs")
```
Comment (sunhay) [inline pkg/util/ecs/common.go]: Trying to clean up some noisy logging in the process agent at the moment. Can we remove this log, and just return the `err`? This gets logged by its upstream caller in (`GetContainers()` in providers.go)

File: pkg/util/ecs/common.go [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+		log.Errorf("decoding task metadata failed - %s", err)
+	}
+	return meta, err
+}
+
+// GetECSContainers returns all containers exposed by the ECS API as plain ECSContainers
+func GetECSContainers() ([]Container, error) {
+	meta, err := GetTaskMetadata()
+	if err != nil || len(meta.Containers) == 0 {
+		log.Errorf("unable to retrieve task metadata")
```
Comment (sunhay) [inline pkg/util/ecs/common.go]: Trying to clean up some noisy logging in the process agent at the moment. Can we remove this log, and just return the `err`? This gets logged by its upstream caller in (`GetContainers()` in providers.go)

Comment (sunhay) [reply pkg/util/ecs/common.go]: For context, if the user does not have any listeners configured, we're setting both `ecs` and `docker` listeners as a default in the process-agent.

It's likely that ECS will not be enabled on most setups, though that means it'll log quite a few errors at the moment. I've copied & modified `providers.go` on our clean up the logging a bit for now but this remains: https://github.com/DataDog/datadog-process-agent/pull/93/files#diff-26a7c4d5ec4700a56067f3a088d2de55R32

**Comment by tmichelet** [comment]
Had a quick look, we're only querying the local metadata, right?

**Comment by hkaj** [comment]
@tmichelet yeah we query the local task metadata.

**Pull request merged**